mantidproject/mantid
# Fix detector 2theta width calculation in SofQWNormalisedPolygon

This fixes the detector angular (2theta) width calculation in `SofQWNormalisedPolygon`. The algorithm was erroneously calculating the angle between the detector center and a point in the center of the detector's bounding box's 'roof', that is, the opening angle of the detector along its 'up' axis. What should be calculated instead is the difference between the 2theta angles of these points.

Further, the algorithm was assuming that all detectors have the same L2 for detector groups. This is not true for instruments like IN5 at the ILL. The 2theta width is now chosen as the range of all 2thetas in a group.

As an additional small improvement  the algorithm now supports sample positions not at the origin.

**Report to:** Björn Fåk & Jacques Ollivier/ILL.

**To test:**

Careful code review. Run the following script both in master and in this branch saving the `sqw` workspace and compare the results:
```python
ws = DirectILLCollectData('ILL/IN4/084446.nxs', FlatBkg='Flat Bkg OFF')

for i in range(ws.getNumberHistograms()):
    Ys = ws.dataY(i)
    Es = ws.dataE(i)
    Ys.fill(0.)
    Es.fill(0.)
    elp = 175
    if i in [71, 271] or i in [9, 209]:
        Ys[elp] = 100.
        Es[elp] = 10.

sqw = DirectILLReduction(ws)
```

On master:
- note the width of the two peaks at energy transfer = ~0meV (ws index 175). One would expect them to occupy ~0.03Å-1
- there are no gaps in the S(Q,W). There should be some at low q and energy transfer close to 0 due to the angular gap between the low angle 'rosace' detectors and the wide angle tubes

On this branch:
- the two peaks should be considerably narrower
- a gap should open at low q between the data from 'rosace' and the wide angle detectors

Fixes #24041.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
